ArrayIndexOutOfBoundsException

I've been getting an ArrayIndexOutOfBoundsException while parsing a response in the `didResolve()` method in `ResponseNormalizer`.  The code is this line:
```
path.remove(path.size() - 1);
```

and path.size() is 0

Using `0.4.0-SNAPSHOT`

```
Caused by: java.lang.ArrayIndexOutOfBoundsException: length=10; index=-1
      at java.util.ArrayList.remove(ArrayList.java:480)
      at com.apollographql.apollo.internal.cache.normalized.ResponseNormalizer.didResolve(ResponseNormalizer.java:52)
      at com.apollographql.apollo.internal.reader.RealResponseReader.didResolve(RealResponseReader.java:194)
      at com.apollographql.apollo.internal.reader.RealResponseReader.readList(RealResponseReader.java:140)
      at com.ibotta.android.ModulesGraphQLQuery$Module$Mapper.map(ModulesGraphQLQuery.java:563)
      at com.ibotta.android.ModulesGraphQLQuery$Data$Mapper$1$1.read(ModulesGraphQLQuery.java:304)
      at com.ibotta.android.ModulesGraphQLQuery$Data$Mapper$1$1.read(ModulesGraphQLQuery.java:301)
      at com.apollographql.apollo.internal.reader.RealResponseReader$ListItemReader.readObject(RealResponseReader.java:251)
      at com.ibotta.android.ModulesGraphQLQuery$Data$Mapper$1.read(ModulesGraphQLQuery.java:301)
      at com.ibotta.android.ModulesGraphQLQuery$Data$Mapper$1.read(ModulesGraphQLQuery.java:298)
      at com.apollographql.apollo.internal.reader.RealResponseReader.readList(RealResponseReader.java:133)
      at com.ibotta.android.ModulesGraphQLQuery$Data$Mapper.map(ModulesGraphQLQuery.java:298)
      at com.ibotta.android.ModulesGraphQLQuery$Data$Mapper.map(ModulesGraphQLQuery.java:293)
      at com.apollographql.apollo.internal.interceptor.HttpResponseBodyParser$1.read(HttpResponseBodyParser.java:57)
      at com.apollographql.apollo.internal.json.ResponseJsonStreamReader.nextObject(ResponseJsonStreamReader.java:86)
      at com.apollographql.apollo.internal.interceptor.HttpResponseBodyParser.parse(HttpResponseBodyParser.java:51)
      at com.apollographql.apollo.internal.interceptor.ApolloParseInterceptor.parse(ApolloParseInterceptor.java:93)
```

And the response looks like this:
```
{
  "data": {
    "modules": [
      {
        "content": [
          {
            "id": "1",
            "info": null
          },
          {
            "id": "2",
            "info": null
          }
        ]
      },
      {
        "content": [
          {
            "id": "3",
            "info": {
              "android": {
                "id": "btn-1a1474fc05e4c7ed",
                "url": null
              },
              "ios": {
                "id": "btn-3e948def5783f2cf",
                "url": null
              }
            }
          },
          {
            "id": "4",
            "info": {
              "android": {
                "id": "btn-1a1474fc05e4c7ed",
                "url": null
              },
              "ios": {
                "id": "btn-3e948def5783f2cf",
                "url": null
              }
            }
          }
        ]
      }
   }
}
```

"info" and all of it's children are nullable fields.   If I remove the "info" field from the query, the response parses just fine.

Could this be related to #552 ?

--- a/apollo-integration/src/test/java/com/apollographql/apollo/ResponseNormalizationTest.java
+++ b/apollo-integration/src/test/java/com/apollographql/apollo/ResponseNormalizationTest.java
@@ -8,6 +8,7 @@ import com.apollographql.apollo.cache.normalized.Record;
 import com.apollographql.apollo.cache.normalized.lru.EvictionPolicy;
 import com.apollographql.apollo.cache.normalized.lru.LruNormalizedCacheFactory;
 import com.apollographql.apollo.exception.ApolloException;
+import com.apollographql.apollo.integration.httpcache.AllPlanetsQuery;
 import com.apollographql.apollo.integration.normalizer.EpisodeHeroNameQuery;
 import com.apollographql.apollo.integration.normalizer.HeroAndFriendsNamesQuery;
 import com.apollographql.apollo.integration.normalizer.HeroAndFriendsNamesWithIDForParentOnlyQuery;
@@ -323,4 +324,24 @@ public class ResponseNormalizationTest {
     assertThat(lukeRecord.field("height(unit:FOOT)")).isEqualTo(BigDecimal.valueOf(5.905512));
   }
 
+  @Test public void list_of_objects_with_null_object() throws Exception {
+    server.enqueue(mockResponse("AllPlanetsListOfObjectWithNullObject.json"));
+
+    AllPlanetsQuery query = new AllPlanetsQuery();
+
+    Response<AllPlanetsQuery.Data> body = apolloClient.query(query).execute();
+    assertThat(body.hasErrors()).isFalse();
+
+    Record record =  normalizedCache
+        .loadRecord("allPlanets(first:300.0).planets.0", CacheHeaders.NONE);
+    assertThat(record.field("filmConnection")).isNull();
+
+    record =  normalizedCache
+        .loadRecord("allPlanets(first:300.0).planets.0.filmConnection", CacheHeaders.NONE);
+    assertThat(record).isNull();
+
+    record =  normalizedCache
+        .loadRecord("allPlanets(first:300.0).planets.1.filmConnection", CacheHeaders.NONE);
+    assertThat(record).isNotNull();
+  }
 }

--- a/apollo-integration/src/test/resources/AllPlanetsListOfObjectWithNullObject.json
+++ b/apollo-integration/src/test/resources/AllPlanetsListOfObjectWithNullObject.json
@@ -1,0 +1,47 @@
+{
+  "data": {
+    "allPlanets": {
+      "__typename": "PlanetsConnection",
+      "planets": [
+        {
+          "__typename": "Planet",
+          "name": "Tatooine",
+          "climates": [
+            "arid"
+          ],
+          "surfaceWater": 1,
+          "filmConnection": null
+        },
+        {
+          "__typename": "Planet",
+          "name": "Alderaan",
+          "climates": [
+            "temperate"
+          ],
+          "surfaceWater": 40,
+          "filmConnection": {
+            "__typename": "PlanetFilmsConnection",
+            "totalCount": 2,
+            "films": [
+              {
+                "__typename": "Film",
+                "title": "A New Hope",
+                "producers": [
+                  "Gary Kurtz",
+                  "Rick McCallum"
+                ]
+              },
+              {
+                "__typename": "Film",
+                "title": "Revenge of the Sith",
+                "producers": [
+                  "Rick McCallum"
+                ]
+              }
+            ]
+          }
+        }
+      ]
+    }
+  }
+}

--- a/apollo-runtime/src/main/java/com/apollographql/apollo/internal/cache/normalized/ResponseNormalizer.java
+++ b/apollo-runtime/src/main/java/com/apollographql/apollo/internal/cache/normalized/ResponseNormalizer.java
@@ -82,10 +82,12 @@ public abstract class ResponseNormalizer<R> implements ResponseReaderShadow<R> {
 
   @Override public void didResolveObject(ResponseField field, Optional<R> objectSource) {
     path = pathStack.pop();
-    Record completedRecord = currentRecordBuilder.build();
-    valueStack.push(new CacheReference(completedRecord.key()));
-    dependentKeys.add(completedRecord.key());
-    recordSet.merge(completedRecord);
+    if (objectSource.isPresent()) {
+      Record completedRecord = currentRecordBuilder.build();
+      valueStack.push(new CacheReference(completedRecord.key()));
+      dependentKeys.add(completedRecord.key());
+      recordSet.merge(completedRecord);
+    }
     currentRecordBuilder = recordStack.pop().toBuilder();
   }
 

--- a/apollo-runtime/src/main/java/com/apollographql/apollo/internal/reader/RealResponseReader.java
+++ b/apollo-runtime/src/main/java/com/apollographql/apollo/internal/reader/RealResponseReader.java
@@ -131,9 +131,9 @@ import java.util.Map;
         Object value = values.get(i);
         if (value != null) {
           T item = (T) listReader.read(new ListItemReader(field, value));
-          readerShadow.didResolveElement(i);
           result.add(item);
         }
+        readerShadow.didResolveElement(i);
       }
       readerShadow.didResolveList(values);
     }

--- a/apollo-runtime/src/main/java/com/apollographql/apollo/internal/reader/RealResponseReader.java
+++ b/apollo-runtime/src/main/java/com/apollographql/apollo/internal/reader/RealResponseReader.java
@@ -109,8 +109,8 @@ import java.util.Map;
     } else {
       parsedValue = (T) objectReader.read(new RealResponseReader(operationVariables, value, fieldValueResolver,
           customTypeAdapters, readerShadow));
-      readerShadow.didResolveObject(field, Optional.fromNullable(value));
     }
+    readerShadow.didResolveObject(field, Optional.fromNullable(value));
     didResolve(field);
     return parsedValue;
   }

